### PR TITLE
[4613][REF][IMP] base_data_import, product_plm_import: refactor _check_field_vals() and add quality_check_categ_id for import

### PR DESCRIPTION
**Pull request by AungKoKoLin1997**
[4613](https://www.quartile.co/web#id=4613&cids=3&menu_id=517&model=project.task&view_type=form)

### CODE REVIEW

File: product_plm_import/wizards/product_plm_import.py [hunk 1]
```diff
@@ -22,6 +22,7 @@
     ["item_type", "Item Type", "char", True],
     ["category", "Category", "char", True],
     ["uom", "Unit of Material", "char", False],
+    ["quality_check_categ_id", "Quality Check Category", "char", False],
```
Comment (yostashiro) [inline product_plm_import/wizards/product_plm_import.py]: Following the request from the client. Never have `_id` in the variable name unless it really is an ID.

```suggestion
    ["acceptance_test_categ", "Acceptance Test Category", "char", False],
```

File: product_plm_import/wizards/product_plm_import.py [hunk 2]
```diff
@@ -69,6 +70,21 @@ def _update_vals_list(self, row_dict, error_list, vals_list):
...
+            row_dict["quality_check_categ_id"] = (
+                quality_check_categ.id if quality_check_categ else False
+            )
+            if not quality_check_categ:
+                error_list.append(
+                    _(
+                        "A non-existent inspection category (%s) is present in the I/F file."
+                    )
+                    % quality_check_categ_code
+                )
```
Comment (yostashiro) [inline product_plm_import/wizards/product_plm_import.py]:
```suggestion
            row_dict["quality_check_categ_id"] = quality_check_categ.id
            if not quality_check_categ:
                error_list.append(
                    _("Quality Check Category (%s) does not exist.", quality_check_categ_code)
                )
```

See this: https://www.odoo.com/documentation/16.0/contributing/development/coding_guidelines.html#use-translation-method-correctly

**Review by yostashiro — CHANGES_REQUESTED** [review]

File: product_plm_import/models/product_plm.py [hunk 3]
```diff
@@ -84,7 +85,25 @@ def _get_uom(self):
...
+                    "error_message": _(
+                        "Quality Check Category (%s) does not exist.",
+                        self.acceptance_test_categ,
+                    ),
+                    "state": "failed",
+                }
+            )
+        return False
+
+    def _create_product(self, quality_check_categ=False):
```
Comment (yostashiro) [inline product_plm_import/models/product_plm.py]:
```suggestion
    def _create_product(self, quality_check_categ):
```

File: product_plm_import/models/product_plm.py [hunk 4]
```diff
@@ -98,6 +117,9 @@ def _create_product(self):
             "detailed_type": mapping.product_type,
             "categ_id": mapping.product_categ_id.id,
             "uom_id": uom.id,
+            "quality_check_categ_id": quality_check_categ.id
+            if quality_check_categ
+            else False,
```
Comment (yostashiro) [inline product_plm_import/models/product_plm.py]:
```suggestion
            "quality_check_categ_id": quality_check_categ.id,
```

File: product_plm_import/models/product_plm.py [hunk 5]
```diff
@@ -84,7 +85,25 @@ def _get_uom(self):
...
+            self.write(
+                {
+                    "error_message": _(
+                        "Quality Check Category (%s) does not exist.",
+                        self.acceptance_test_categ,
+                    ),
+                    "state": "failed",
+                }
+            )
+        return False
```
Comment (yostashiro) [inline product_plm_import/models/product_plm.py]: To make it slightly more concise, adding the missing `self.ensure_one()`.
```suggestion
        self.ensure_one()
        quality_check_categ = self.env["quality.check.category"].browse()
        test_categ = self.acceptance_test_categ
        if test_categ:
            quality_check_categ = self.env["quality.check.category"].search(
                [("code", "=", test_categ)], limit=1
            )
            if not quality_check_categ:
                self.error_message = _(
                    "Quality Check Category (%s) does not exist.", test_categ,
                )
                self.state = "failed"
        return quality_check_categ
```

**Review by yostashiro — COMMENTED** [review]
I actually don't feel comfortable with the proposed changes (doesn't current proposal override the error message?).

What I'm thinking is to improve the check mechanism at the low level:

- Cut this part into a method (`_validate_field_value()` or something): https://github.com/qrtl/axls-custom/blob/09dec36ecdd3ebcd839e9eb59370c59c2f6ee727/base_data_import/wizards/data_import.py#L121-L138
- Extend this method in `product_plm_import` to add the checking on "Acceptance Test Category", and update `row_dict` with `quality_check_categ_id` if there is a corresponding record
- Similarly, check against the product (which is currently done in `_update_vals_list`) in the extended method

The idea is to complete all the field checks within `_check_field_vals()`.  Does this make sense?

**Review by yostashiro — APPROVED** [review]
Code review.

**Review by kanda999 — APPROVED** [review]
Code and Functional review.

**Review by yostashiro — APPROVED** [review]
👍

**Review by kanda999 — APPROVED** [review]
Functional test is working.

**Pull request merged**